### ISSUE TITLE
Filter function not recognised

### ISSUE THREAD

**Issue by BenLaneUK2004** [withastro/astro#10771]
### Astro Info

```block
astro : The term 'astro' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, 
or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ astro info
+ ~~~~~
    + CategoryInfo          : ObjectNotFound: (astro:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

### Describe the Bug

I'm following the blog tutorial, and I have got to the tags section. There is a problem with this piece of code from the tutorial:
`const filteredPosts = posts.filter((post) => post.frontmatter.tags?.includes(tag));`

This is the VScode error: 

> Property 'filter' does not exist on type 'never'.ts(2339)

This is the astro development server error: 

> Cannot read properties of undefined (reading 'filter')

I copied the code exactly, so I'm wondering if I did something wrong or if it is a bug.

### What's the expected result?

There should be no error, as it is verbatim code from the offcial docs

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-cb3vcc?file=src%2Fpages%2Findex.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Princesseuh** [comment]
Hello! You are using this code on the wrong page of your website. If you look at the original example, you'll notice that this snippet is part of a dynamic route at `/src/pages/tags/[tag].astro`. 

I suggest rereading the section: https://docs.astro.build/en/tutorial/5-astro-api/2/#advanced-javascript-generate-pages-from-existing-tags

**Comment by BenLaneUK2004** [comment]
@Princesseuh That is where it is, I just chucked it on that page in the stackblitz example. Here is code in my work: 
![image](https://github.com/withastro/astro/assets/126700601/ff5e1467-1a2f-48d7-8a8c-a98717466b52)

**Comment by Princesseuh** [comment]
You do not pass the posts as the props of your pages, as such there's never any pages being passed.

**Comment by albano-a** [comment]
The issue is still there. I was following the tutorial, and got the same problem

**Comment by Beckalina** [comment]
I'm having the same problem, my code is the same as @BenLaneUK2004 's example, but I don't understand your explanation @Princesseuh , where do I need to pass the posts? This is all that is added in the tutorial and they say it should work so maybe the tutorial has a step missing? https://docs.astro.build/en/tutorial/5-astro-api/2/

**Comment by Beckalina** [comment]
Ah - from scrolling further down in the tutorial, a later example shows the return value of the getStaticPaths function as follows:
